LFS-250 / LFS-284: Vocabulary term suggestions ignore the current vocabulary / Vocabulary widgets should use locally installed vocabularies

OrdiNeu opened the pull request:
[LFS-250 Jira link](https://phenotips.atlassian.net/browse/LFS-250) / [LFS-284 Jira link](https://phenotips.atlassian.net/browse/LFS-284)

To test: Create a new **Demographics** form. The vocabulary used will now rely on the locally installed versions, which means that e.g. the Co-morbidities question will not work by default. To get them to work, install the Human Phenotypes Ontology from the Vocabularies admin section on the left. Similarly, the Chemotherapy Agent question in **Chemotherapy** will require the CHEBI vocabulary to be installed.

Note that none of the vocabulary questions will work by default anymore.

OrdiNeu committed:
LFS-250: Vocabulary widgets should use locally installed vocabularies
Create a new servlet for parsing through references in VocabularyTerms,
and add the ability to specify filters/sorting on the SearchTermServlet

OrdiNeu committed:
LFS-250: Vocabulary Term suggestions ignore the current vocabulary
Attempt to reduce the number of requests to/from the server by including
information on which lfs:VocabularyTerms have children during Vocab
browsing.

OrdiNeu committed:
LFS-250: Vocabulary term suggestions ignore the current vocabulary
Prevent re-downloading of browser branches when we don't yet need to
know their children.
Revert search.jsx to its original

OrdiNeu committed:
[misc] Remove wip SolR code.

sashaandjic commented:
Works well. 
I would like to be able to select Incomitant strabismus from the Related terms
![image](https://user-images.githubusercontent.com/23243244/74973968-6894e180-53f2-11ea-852e-a1bb80a9b517.png)

marta- commented:
> I would like to be able to select Incomitant strabismus from the Related terms

Please create a task @sashaandjic . Thanks!

sashaandjic commented:
When using CHEBI there is a delay displaying info box as well as related term box.
![image](https://user-images.githubusercontent.com/23243244/75060877-9ee36700-54ad-11ea-83a5-a60fa889ce62.png)

sashaandjic commented:
It would be nice to have a focus back at the new line after selecting a vocabulaty term. This behaviour is already present for multi value entries without lookup.
![image](https://user-images.githubusercontent.com/23243244/75061885-a86dce80-54af-11ea-91a7-754458349cac.png)

OrdiNeu committed:
LFS-250: Vocabulary term suggestions ignore the current vocabulary
Extend the suggestion search to match on a partially entered term.

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Refocus on the select input after a term is selected.

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Massively speedup info searches by restricting the parent-finding search
to only children of the term's vocabulary

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Move the vocabulary description tooltip closer to the avatar in chrome.

OrdiNeu committed:
LFS-250: Vocabulary widgets should use locally installed vocabularies
Styling fix

sashaandjic commented:
Upon successful selection of HPO term Melasma, the Melasma tooltip gets displayed on the blank line below the term

marta- commented:
> Upon successful selection of HPO term Melasma, the Melasma suggestion gets displayed on the blank line below the term even though I haven't typed anything.

@OrdiNeu I think this may be triggered by commit c28c94c0ba2a5d, can you look into it?

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -370,27 +374,29 @@ class VocabularyQuery extends React.Component {
-                    this.anchorEl.value = element["name"];
+                    this.props.onClick(element["identifier"], name);
+                    this.anchorEl.value = name;
+                    this.anchorEl.select();
```
This should probably be done after the value of the input is cleared, e.g. inside closeDialog. Otherwise it seems to cause another lookup and display a list of suggestions.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/util.jsx:
```diff
@@ -17,8 +17,8 @@
-// Do not place an ending / in the URL below
-export const REST_URL = "https://services.phenotips.org/rest/vocabularies";
+// Ensure there is an endling / in the URL below
+export const REST_URL = window.location.origin + "/Vocabularies/";
```
`window.location.origin` shouldn't be needed. Moreover, it might break things if, for example, we define a different [base url](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base) (we don't at this time, but let's keep our options open).

```suggestion
export const REST_URL = "/Vocabularies/";
```

marta- replied:
Turns out this won't actually work. We'll try to accomodate the mentioned usecase when we actually encounter it. The current code works just right for now.

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Move the text refocus until after the suggestion box is cleared.

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermInfoServlet.java:
```diff
@@ -0,0 +1,233 @@
+    {
+        Resource vocab = request.getResource().getParent();
+        ResourceResolver resolver = request.getResourceResolver();
+        String parentpath = this.findParentVocabulary(request.getResource()).getPath();
```
`vocab` and `parentpath` are obtained in different ways. Right now they still work in the same way, but the assumption that all terms are direct children of a vocabulary node may no longer hold in the future. For safety, I would do `vocab = findParentVocabulary`, and `parentPath` should then be ` = vocab.getPath()`.

sdumitriu replied:
Why not `parentPath`? Also in all the other places in this file and the next where the uncamelcased name is used.

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermInfoServlet.java:
```diff
@@ -0,0 +1,233 @@
+    @Override
+    public void doGet(final SlingHttpServletRequest request, final SlingHttpServletResponse response) throws IOException
+    {
+        Resource vocab = request.getResource().getParent();
```
Why not `vocabulary`?

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermInfoServlet.java:
```diff
@@ -0,0 +1,233 @@
+     * @param parentpath the location of the vocabulary whose children we're searching
+     * @return An iterator over the resource's children
+     */
+    private Iterator<Resource> getVocabularyTermChildren(JsonObject resource, ResourceResolver resolver,
```
This class only deals with vocabulary terms, and the other methods don't use `VocabularyTerm` in their name, so for consistency and simplicity:

```suggestion
    private Iterator<Resource> getChildren(JsonObject resource, ResourceResolver resolver,
```

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermSearchServlet.java:
```diff
@@ -61,86 +68,183 @@ public void doGet(final SlingHttpServletRequest request, final SlingHttpServletR
     {
```
This method is getting kind of long, can it be split?

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermSearchServlet.java:
```diff
@@ -61,86 +68,183 @@ public void doGet(final SlingHttpServletRequest request, final SlingHttpServletR
+        // The writer doesn't need to be explicitly closed since the auto-closed jsonGen will also close the writer
+        response.setContentType("application/json");
+        final Writer out = response.getWriter();
+        try (JsonGenerator jsonGen = Json.createGenerator(out)) {
```
This block is getting duplicated too often. We should write a common function for it. But maybe not right now, something for a cleanup sprint.

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermInfoServlet.java:
```diff
@@ -0,0 +1,233 @@
+    {
+        // Check to see if this resource has children
+        String oakQuery = String.format(
+            "SELECT * FROM [lfs:VocabularyTerm] AS a WHERE isdescendantnode(a, '%s') AND a.parents = '%s'",
```
Add an order by label.

sdumitriu reviewed (changes requested):
If HP isn't indexed first, the error is a bit uninformative and unintuitive: 
- `Error: Thesaurus lookup failed with code 404` displayed at the bottom of the window, should be closer to the input
- the message should be more informative, suggesting that `the needed vocabulary is not available, contact an admin to install it`
- the field is stuck in the "loading..." state

marta- commented:
> If HP isn't indexed first, the error is a bit uninformative and unintuitive:

@sdumitriu LFS-347, we discussed it and decided it was out of scope.

marta- commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermInfoServlet.java:
```diff
@@ -0,0 +1,233 @@
+
+    private static final String CHILDREN_PROPERTY = "lfs:children";
+
+    /* During tests, copying over every child tends to bloat the response, so we only copy a subset of the data */
```
"According to manual tests"?

OrdiNeu committed:
[misc] Remove unused SolR reconfiguration

OrdiNeu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermInfoServlet.java:
```diff
@@ -0,0 +1,233 @@
+
+        // Copy over our children to lfs:Children
+        JsonArrayBuilder childrenBuilder = Json.createArrayBuilder();
+        while (children.hasNext()) {
```
Originally I only tracked whether it had children, but adding this much info lets me skip a lot of back-and-forth requests when building the browser tree.

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Apply codestyle improvement suggestions by @sdumitriu

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Add an order by label when finding children of a vocab term.

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Refactor doGet in VocabularyTermSearchServlet into smaller functions.

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Made finding children of parent resources an optional parameter

OrdiNeu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyTermInfoServlet.java:
```diff
@@ -0,0 +1,233 @@
+                    builder.add(value);
+                } else {
+                    JsonObject linkedObject = linkedValue.adaptTo(JsonObject.class);
+                    builder.add(populateChildren(linkedObject, resolver, parentpath));
```
For some reason I thought it was necessary, but it apparently is not; I've set it up so you can pass `includeChildren=true` if necessary, but currently none of the code requires it.

OrdiNeu committed:
[misc] minor codestyle change

marta- reviewed (approved):
No more complaints on my side, other than the UI/UX fixes which are out of scope and for which we created tickets.

sashaandjic commented:
Problem found: the vocabulary entries do not get saved . See the blank "Co-morbidities..."  for http://172.20.4.63:8080/content.html/Forms/c9bd1247-a287-481c-9c1c-fbec1df740d1 

![image](https://user-images.githubusercontent.com/23243244/75491775-7ca58980-5984-11ea-8cf2-7925d2c796ca.png)
![image](https://user-images.githubusercontent.com/23243244/75492003-d443f500-5984-11ea-93ec-48c3551ac30c.png)

OrdiNeu committed:
LFS-284: Vocabulary widgets should use locally installed vocabularies
Bugfix on vocab answers not loading properly.

OrdiNeu commented:
> Problem found: the vocabulary entries do not get saved . See the blank "Co-morbidities..." for http://172.20.4.63:8080/content.html/Forms/c9bd1247-a287-481c-9c1c-fbec1df740d1
@sashaandjic I just fixed it in commit `4b13619`.

sashaandjic reviewed (approved):
tested
